### PR TITLE
Fix Transpose bug in Inequality Regression

### LINKED ISSUE #303: [QUESTION]: how to handle mixed continuous-categorical data?
Suppose you have experimental activity data, say pIC50's at some target for a set of compounds, but the experiment was only able to measure activity in the interval [5, 9.5], below and above which the data are reported as '< 5' or '> 9.5', respectively. This means of course that some data are continuous, other data categorical.

I saw a 'bounded_mse' option to handle such data in chemprop, but - please correct me if I am wrong - when I looked at the code, it seemed to me that it only strips the qualifier, i.e. any data with '< 5' becomes 5, and '> 9.5' becomes 9.5.
Then, if I understand correctly, the model will always output continuous values in regression mode, so if it predicts a pIC50 of 5, 4, 3, ..., -1..., for compounds that have '< 5' as the original pIC50, it will simply add to the mse a corresponding squared residual using the fixed un-censored value (5): (5-5)^2, (5-4)^2, (5-3)^2, ..., (5-(-1))^2, ...,?
But in fact the model is 'right' in predicting a value below 5 for a compound that has '< 5', so why should it add a relatively large squared residual contribution to the mse for such a prediction?

This case is quite frequent in practice, especially with ADMET data, and censored values are often a substantial fraction of the total data points, so the information they bring is valuable and should be used rather than, e.g. filtered out. And indeed, various approaches to handle such cases are described in the literature, most notably censored regression methods (such as Tobit), which were also implemented in some packages (e.g. https://github.com/fabsig/KTBoost).

Is this aspect already addressed by chemprop by some option I missed? And if not, is there any plan to include it in future releases?

Thanks

### PR DESCRIPTION
**Pull request by cjmcgill**
## Description
Metric evaluation in `chemprop/train/evaluate.py` has a step where the targets and predictions made by the model are transposed from a (data, tasks) shape into a (tasks, data) shape. This is done so that the evaluation can skip values missing in the targets during metric evaluation. The metrics for use with inequality also take in additional inputs marking the locations of inequality targets (`lt_targets`, `gt_targets`). These additional inputs were not transposed. The result is that the correct boolean indicators of inequality are not being used in the metrics. The possible situations that I see arising:

- Multitask regression fails outright with the below error or similar, because the number of tasks isn't the same as the number of data being evaluated, so the un-transposed values cause errors in metric calculation.

```
File ".../chemprop/chemprop/train/metrics.py", line 160, in bounded_mse
    np.logical_and(np.greater(preds, targets),gt_targets),
ValueError: operands could not be broadcast together with shapes (2,) (3,) 
```

- Single task. The operation takes the `gt_target` or `lt_target` for a single datapoint and applies it across all data for a single task. The first datapoint being used for the first task, second datapoint for the second task, etc.

## Questions
The transposition in the evaluation method is awkward and difficult to read if you don't already know what's going on. It may be better to restructure the whole section so the transposition is avoided or clearer.

## Relevant issues
#303 (secondary question) notes the error when multitask inequality regressions are used. This PR resolves #303 .

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by degraff-roivant** [comment]
is there a reason we're doing evaluation with native python lists instead of numpy arrays or tensors? It's not reducing a depedency and transposing lists (as you've mentioned) is pretty verbose

**Comment by cjmcgill** [comment]
@davidegraff the reason it's not a normal transpose is because we knock out missing values in some tasks, so the lengths wouldn't be uniform. I think that you could do it so that it's much clearer, but I don't see a way to keep it in a single np array after masking out values.

If I were rewriting the whole thing, I think I'd get rid of the transpose entirely. We loop through the tasks down at L64 anyway (because of the uneven task length). I would put the masking step inside that loop. Maybe I should just go ahead and do this.

**Comment by cjmcgill** [comment]
Transpose removed. A change I have wanted to make for a while and finally had the chance.

File: chemprop/train/evaluate.py [hunk 1]
```diff
@@ -31,68 +35,94 @@ def evaluate_predictions(preds: List[List[float]],
...
-    # valid_preds and valid_targets have shape (num_tasks, data_size)
-    if dataset_type != 'spectra':
-        valid_preds = [[] for _ in range(num_tasks)]
-        valid_targets = [[] for _ in range(num_tasks)]
-        for i in range(num_tasks):
-            for j in range(len(preds)):
-                if targets[j][i] is not None:  # Skip those without targets
-                    valid_preds[i].append(preds[j][i])
-                    valid_targets[i].append(targets[j][i])
+    # Convert preds and targets to np arrays
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:43]:
```suggestion
```

File: chemprop/train/evaluate.py [hunk 2]
```diff
@@ -31,68 +35,94 @@ def evaluate_predictions(preds: List[List[float]],
...
-                    info('Warning: Found a task with predictions all 0s or all 1s')
-
-                if nan:
-                    for metric in metrics:
-                        results[metric].append(float('nan'))
+            if dataset_type == "classification":
+                return_nan = False
+                if all(target == 0 for target in task_targets) or all(
+                    target == 1 for target in task_targets
+                ):
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:72]:
```suggestion
                if (task_targets == 0).all() or (task_targets == 1).all():
```

File: chemprop/train/evaluate.py [hunk 3]
```diff
@@ -31,68 +35,94 @@ def evaluate_predictions(preds: List[List[float]],
...
+                    return_nan = True
+                    info("Warning: Found a task with targets all 0s or all 1s")
+                if all(pred == 0 for pred in task_preds) or all(
+                    pred == 1 for pred in task_preds
+                ):
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:77]:
```suggestion
                if (task_preds == 0).all() or (task_preds == 1).all():
```

File: chemprop/train/evaluate.py [hunk 4]
```diff
@@ -31,68 +35,94 @@ def evaluate_predictions(preds: List[List[float]],
...
-    if dataset_type == 'spectra':
-        for metric, metric_func in metric_to_func.items():
-            results[metric].append(metric_func(preds, targets))
+    if dataset_type == "spectra":
+        for metric_label, metric_func in metric_to_func.items():
+            results[metric_label].append(metric_func(preds, targets))
     else:
         for i in range(num_tasks):
+            # remove targets/predictions from evaluation that have no target values
+            task_mask = ~np.isnan(targets[:, i])
```
[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/train/evaluate.py:59]: would probably want to change the next two lines as well

[the remark this answers is not in the document]
Comment (cjmcgill) [reply chemprop/train/evaluate.py:58]: I could see defining `mask` earlier and taking a slice of it like we do for targets and preds, `task_mask = mask[:, i]`. That would have a nice analogous structure.

File: chemprop/train/evaluate.py [hunk 5]
```diff
@@ -31,68 +35,89 @@ def evaluate_predictions(preds: List[List[float]],
...
+                    )
                 else:
-                    results[metric].append(metric_func(valid_targets[i], valid_preds[i]))
+                    results[metric_label].append(
+                        metric_func(task_targets, task_preds)
+                    )
 
     results = dict(results)
 
     return results
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:109]:
```suggestion
    return dict(results)
```

File: chemprop/train/evaluate.py [hunk 6]
```diff
@@ -2,20 +2,24 @@
...
-                         num_tasks: int,
-                         metrics: List[str],
-                         dataset_type: str,
-                         gt_targets: List[List[bool]] = None,
-                         lt_targets: List[List[bool]] = None,
-                         logger: logging.Logger = None) -> Dict[str, List[float]]:
+def evaluate_predictions(
+    preds: List[List[float]],
+    targets: List[List[float]],
+    num_tasks: int,
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:16]: is this argument still necessary? should be inferable from the shape of both `preds` and `targets`

File: chemprop/train/evaluate.py [hunk 7]
```diff
@@ -31,68 +35,89 @@ def evaluate_predictions(preds: List[List[float]],
...
+                    info("Warning: Found a task with predictions all 0s or all 1s")
+
+                if return_nan:
+                    for metric_label in metrics:
+                        results[metric_label].append(float("nan"))
                     continue
 
-            if len(valid_targets[i]) == 0:
+            if len(task_targets) == 0:
                 continue
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:82]: move up towards top of loop so we avoid some potentially unnecessary logic

File: chemprop/train/evaluate.py [hunk 8]
```diff
@@ -31,68 +35,89 @@ def evaluate_predictions(preds: List[List[float]],
...
+                            task_lt_targets,
+                        )
+                    )
                 else:
-                    results[metric].append(metric_func(valid_targets[i], valid_preds[i]))
+                    results[metric_label].append(
+                        metric_func(task_targets, task_preds)
+                    )
 
     results = dict(results)
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:107]:
```suggestion
    return results
```
`results` is already of type `dict` (a subclass, specifically)

File: chemprop/train/evaluate.py [hunk 9]
```diff
@@ -31,68 +35,89 @@ def evaluate_predictions(preds: List[List[float]],
...
+        lt_targets = np.array(lt_targets, dtype=bool)
 
     # Compute metric. Spectra loss calculated for all tasks together, others calculated for tasks individually.
     results = defaultdict(list)
-    if dataset_type == 'spectra':
-        for metric, metric_func in metric_to_func.items():
-            results[metric].append(metric_func(preds, targets))
+    if dataset_type == "spectra":
+        for metric_label, metric_func in metric_to_func.items():
+            results[metric_label].append(metric_func(preds, targets))
```
Comment (davidegraff) [inline chemprop/train/evaluate.py:54]:
```suggestion
        results = {metric_label: metric_func(preds, targets) for metric_label, metric_func in metric_to_func.items()}
```
this new suggestion isn't right, but it seems like the original code produces a value of `results: dict[str, [float]]`. I.e., that each value was a singleton list, is that right?

Comment (davidegraff) [reply chemprop/train/evaluate.py:54]: we can also return right after this block and de-indent the remaining code to give it a little more breathing room

**Review by davidegraff — COMMENTED** [review]
mostly questions here

**Comment by oscarwumit** [comment]
Any updates to this PR?

**Comment by kevingreenman** [comment]
@cjmcgill thank you! I agree with reverting back to the minimal solution in the interest of getting this merged. It looks like one of the tests is now failing though

**Comment by cjmcgill** [comment]
@kevingreenman fixing the bug changed the output value for the test because the inequality evaluation wasn't being applied correctly. Should be good to review now.

**Review by kevingreenman — APPROVED** [review]

**Pull request merged**